test-data/unit/check-based-intersection.test

```diff
@@ -0,0 +1,340 @@
+x.foo(None)  # E: No overload variant of "foo" of "__main__.A & __main__.B" matches argument type "None"  [call-overload] \
+             # N: Possible overload variants: \
+             # N:     def foo(self) -> None \
+             # N:     def foo(self, x: int) -> None \
+             # N:     def foo(self, x: str) -> None
+[builtins fixtures/tuple.pyi]
+
+
+[case testGenericFunction]
+from __future__ import annotations
+
+from typing import TypeVar
+
+T = TypeVar('T')
+U = TypeVar('U')
+
+def foo(x: T, y: U) -> T & U: pass
+a: int
+b: object
+reveal_type(foo(a, b))  # N: Revealed type is "int"
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionInferenceWithTypeVarValues]
+from __future__ import annotations
+from typing import TypeVar
+AnyStr = TypeVar('AnyStr', bytes, str)
+class A: pass
+strA: str & A
+def f(x: AnyStr & A, *a: AnyStr): pass
+f(strA)
+f(strA, 'bar')
+f(strA, b'bar') # E: Value of type variable "AnyStr" of "f" cannot be "str | bytes"  [type-var]
+[builtins fixtures/primitives.pyi]
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionTwoPassInference-skip]
+from __future__ import annotations
+from typing import TypeVar
+T = TypeVar('T')
+U = TypeVar('U')
+def j(x: list[T] & list[U], y: list[T]) -> list[U]: pass
+
+a = [1]
+b = ['b']
+# We could infer: Since list[str] <: list[T], we must have T = str.
+# Then since list[int] <: Union[list[str], list[U]], and list[int] is
+# not a subtype of list[str], we must have U = int.
+# This is not currently implemented.
+j(a, b)
+[builtins fixtures/list.pyi]
+[builtins fixtures/tuple.pyi]
+
+
+[case testIntersectionContext]
+from __future__ import annotations
+
+from typing import TypeVar
+
+class A: pass
+class B: pass
+
+T = TypeVar('T')
+def f() -> list[T]: pass
+d1: list[int] & A = f()  # E: Incompatible types in assignment (expression has type "list[int]", variable has type "list[int] & A")  [assignment]
+d2: A & B = f() # E: Incompatible types in assignment (expression has type "list[Never]", variable has type "A & B")  [assignment]
+[builtins fixtures/list.pyi]
+[builtins fixtures/tuple.pyi]
+
+
+[case testGenericFunctionSubtypingWithIntersections]
+# flags: --disallow-redefinition
+from __future__ import annotations
+from typing import TypeVar
+T = TypeVar('T')
+S = TypeVar('S')
+def k1(x: int, y: list[T]) -> list[T & int]: pass
+def k2(x: S, y: list[T]) -> list[T & int]: pass
+a = k2
+a = k1 # E: Incompatible types in assignment (expression has type "[T] (int, list[T]) -> list[T & int]", variable has type "[S, T] (S, list[T]) -> list[T & int]")  [assignment]
+b = k1
+b = k2
+[builtins fixtures/list.pyi]
+[builtins fixtures/tuple.pyi]
+
+
+[case testEqual-xfail]
+# flags: --disallow-redefinition
+from __future__ import annotations
+from typing import Generic
+
+T = TypeVar("T")
+class A: pass
+class B: pass
+class G(Generic[T]): pass
+
+l1: G[A] & G[B]
+l2: G[A & B]
+
+l1 = l2
+l2 = l1
+[builtins fixtures/tuple.pyi]
+
+
+[case testOverloadIsIntersection-xfail]
```

Comment:
why are these xfail? are there todos/issues for these?

Reply:
Yes, all the xfails are mentioned in the OP